pkg/controllers/sync/dispatch/retain.go

```diff
@@ -251,6 +263,231 @@ func retainPersistentVolumeClaimFields(
 	return nil
 }
 
+func retainPodFields(desiredObj, clusterObj *unstructured.Unstructured) error {
+	// A general guideline is to always drop and retain fields that are unable to be set by the user and are managed by
+	// the Kubernetes control plane instead. ephemeralContainers falls into this cateogry.
+	if err := copyUnstructuredField(clusterObj, desiredObj, "spec", "ephemeralContainers"); err != nil {
+		return err
+	}
+
+	// The following fields are fields that can be explicitly set by the user, but are defaulted by the Kubernetes
+	// control plane (after creation) if left unset. For these fields, we retain the defaulted values in clusterObj if
+	// the field was not explicitly set in desiredObj. Otherwise, we respect the user's choice.
+	if serviceAccountName, exists, err := unstructured.NestedString(desiredObj.Object, "spec", "serviceAccountName"); err == nil &&
+		(!exists || len(serviceAccountName) == 0) {
+		if err := copyUnstructuredField(clusterObj, desiredObj, "spec", "serviceAccountName"); err != nil {
+			return err
+		}
+	}
+
+	if serviceAccount, exists, err := unstructured.NestedString(desiredObj.Object, "spec", "serviceAccount"); err == nil &&
+		(!exists || len(serviceAccount) == 0) {
+		if err := copyUnstructuredField(clusterObj, desiredObj, "spec", "serviceAccount"); err != nil {
+			return err
+		}
+	}
+
+	if nodeName, exists, err := unstructured.NestedString(desiredObj.Object, "spec", "nodeName"); err == nil &&
+		(!exists || len(nodeName) == 0) {
+		if err := copyUnstructuredField(clusterObj, desiredObj, "spec", "nodeName"); err != nil {
+			return err
+		}
+	}
+
+	if priority, exists, err := unstructured.NestedFieldNoCopy(desiredObj.Object, "spec", "priority"); err == nil &&
+		(!exists || priority == nil) {
+		if err := copyUnstructuredField(clusterObj, desiredObj, "spec", "priority"); err != nil {
+			return err
+		}
+	}
+
+	if _, _, exists := findServiceAccountVolume(desiredObj); !exists {
+		if volume, idx, exists := findServiceAccountVolume(clusterObj); exists {
+			// If the service account volume exists in clusterObj but not in the desiredObj, it was injected by the
+			// service account admission plugin. We retain the service account volume at the same index in desiredObj
+			// (if we do not preserve the ordering of the volumes slice, the update will fail).
+			desiredVolumes, _, _ := unstructured.NestedSlice(desiredObj.Object, "spec", "volumes") // this is a deepcopy
+			if len(desiredVolumes) < idx {
+				return fmt.Errorf("failed to copy service account volume, slice length mismatch")
+			}
+
+			desiredVolumes = append(desiredVolumes, nil)
+			copy(desiredVolumes[idx+1:], desiredVolumes[idx:])
+			desiredVolumes[idx] = volume
+
+			if err := unstructured.SetNestedSlice(desiredObj.Object, desiredVolumes, "spec", "volumes"); err != nil {
+				return err
+			}
+		}
+	}
+
+	desiredContainers, _, _ := unstructured.NestedSlice(desiredObj.Object, "spec", "containers") // this is a deepcopy
```

Comment:
Why do we not need to handle err here?

Reply:
1. when err is returned, desiredContainers will be `nil` which is handled by `retainContainerFields`. 
2. `unstructured.NestedSlice` returns err when the value for the given field is not a `[]interface{}`, implying that the container field in the pod is corrupted for some reason - in that case, we simply ignore the field (which brings us back to 1)